Rework build, replace xgo with golang-crossbuild

igor-hnizdo opened the pull request:
xgo is deprecated. Let's replace it with Elastic's Golang-crossbuild.

It is better documented and more maintained.

https://github.com/elastic/golang-crossbuild

That docker conveniently also has lipo, which will make the fat binary.

igor-hnizdo commented on release/macos/Makefile:
```diff
@@ -16,8 +16,25 @@ clean:
+		docker.elastic.co/beats-dev/golang-crossbuild:1.19-darwin-arm64-debian10 \
+		--build-cmd "go build -o release/macos/build/trezord-arm64" \
+		-p "darwin/arm64"
+	docker run -it --rm \
```
on mac, there is `lipo` for that; but it's not in docker :) but the actual fat binary seems to be simple, and there is a tool that helps with that

this way, we can just make one binary for all macos and not two separate binaries

igor-hnizdo replied:
Ah nevermind, the elastic thing has lipo included (through osxcross). So I put it there instead of the weird un-maintaned tool.

tsusanka reviewed (commented):
Is this supposed to work no arm macs? I am getting errors for both `darwin-debian10` and `darwin-arm64-debian10` images.

<details>

```

Cleaning...
rm -rf build
Building with xgo ...
mkdir -p build
docker run -it --rm \
		-v /Users/tomassusanka/go/src/github.com/trezor/trezord-go:/trezord \
		-w /trezord \
		-e CGO_ENABLED=1 \
		docker.elastic.co/beats-dev/golang-crossbuild:1.19-darwin-debian10 \
		--build-cmd "go build -o release/macos/build/trezord-amd64" \
		-p "darwin/amd64"
>> Building using: cmd='go build -o release/macos/build/trezord-amd64', env=[CC=o64-clang, CXX=o64-clang++, GOARCH=amd64, GOARM=, GOOS=darwin, PLATFORM_ID=darwin-amd64]
go: downloading gopkg.in/natefinch/lumberjack.v2 v2.0.0
go: downloading github.com/gorilla/handlers v1.5.1
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/gorilla/csrf v1.7.0
go: downloading github.com/felixge/httpsnoop v1.0.1
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/gorilla/securecookie v1.1.1
# github.com/trezor/trezord-go/usb/lowlevel/libusb
cgo-gcc-prolog:1186:2: warning: 'libusb_set_debug' is deprecated [-Wdeprecated-declarations]
PLATFORM  = macos
usb/lowlevel/libusb/c/libusb.h:1351:1: note: 'libusb_set_debug' has been explicitly marked deprecated here
usb/lowlevel/libusb/c/libusb.h:70:50: note: expanded from macro 'LIBUSB_DEPRECATED_FOR'
docker run -it --rm \
		-v /Users/tomassusanka/go/src/github.com/trezor/trezord-go:/trezord \
		-w /trezord \
		-e CGO_ENABLED=1 \
		docker.elastic.co/beats-dev/golang-crossbuild:1.19-darwin-arm64-debian10 \
		--build-cmd "go build -o release/macos/build/trezord-arm64" \
		-p "darwin/arm64"
>> Building using: cmd='go build -o release/macos/build/trezord-arm64', env=[CC=oa64-clang, CXX=oa64-clang++, GOARCH=arm64, GOARM=, GOOS=darwin, PLATFORM_ID=darwin-arm64]
go: downloading gopkg.in/natefinch/lumberjack.v2 v2.0.0
go: downloading github.com/gorilla/handlers v1.5.1
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/gorilla/csrf v1.7.0
go: downloading github.com/felixge/httpsnoop v1.0.1
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/gorilla/securecookie v1.1.1
# github.com/trezor/trezord-go/usb/lowlevel/libusb
cgo-gcc-prolog:1186:2: warning: 'libusb_set_debug' is deprecated [-Wdeprecated-declarations]
usb/lowlevel/libusb/c/libusb.h:1351:1: note: 'libusb_set_debug' has been explicitly marked deprecated here
usb/lowlevel/libusb/c/libusb.h:70:50: note: expanded from macro 'LIBUSB_DEPRECATED_FOR'
docker run -it --rm \
		-v /Users/tomassusanka/go/src/github.com/trezor/trezord-go:/trezord \
		-w /trezord \
		-e CGO_ENABLED=1 \
		--entrypoint /trezord/release/macos/build-fat.sh \
		docker.elastic.co/beats-dev/golang-crossbuild:1.19-darwin-arm64-debian10
+ /usr/local/osxcross/bin/lipo -create release/macos/build/trezord-arm64 release/macos/build/trezord-amd64 -output release/macos/build/trezord
cp ../../VERSION build
Preparing docker image ...
docker build -t trezord-go-build-env-macos .
[+] Building 1.3s (3/3) FINISHED
 => [internal] load build definition from Dockerfile                                            0.0s
 => => transferring dockerfile: 37B                                                             0.0s
 => [internal] load .dockerignore                                                               0.0s
 => => transferring context: 2B                                                                 0.0s
 => ERROR [internal] load metadata for docker.io/library/fedora:25                              1.2s
------
 > [internal] load metadata for docker.io/library/fedora:25:
------
failed to solve with frontend dockerfile.v0: failed to create LLB definition: no match for platform in manifest sha256:322cb01bbca26972c98051bacd3ab8555cec059496d64d35ee78b15de9ea0d06: not found
make: *** [.docker-image] Error 1
~/g/s/g/t/t/r/macos[2]>vi Makefile
~/g/s/g/t/t/r/macos>make all
Cleaning...
rm -rf build
Building with xgo ...
mkdir -p build
docker run -it --rm \
		-v /Users/tomassusanka/go/src/github.com/trezor/trezord-go:/trezord \
		-w /trezord \
		-e CGO_ENABLED=1 \
		docker.elastic.co/beats-dev/golang-crossbuild:1.19-darwin-arm64-debian10 \
		--build-cmd "go build -o release/macos/build/trezord-arm64" \
		-p "darwin/arm64"
>> Building using: cmd='go build -o release/macos/build/trezord-arm64', env=[CC=oa64-clang, CXX=oa64-clang++, GOARCH=arm64, GOARM=, GOOS=darwin, PLATFORM_ID=darwin-arm64]
go: downloading gopkg.in/natefinch/lumberjack.v2 v2.0.0
go: downloading github.com/gorilla/handlers v1.5.1
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/gorilla/csrf v1.7.0
go: downloading github.com/felixge/httpsnoop v1.0.1
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/gorilla/securecookie v1.1.1
# github.com/trezor/trezord-go/usb/lowlevel/libusb
cgo-gcc-prolog:1186:2: warning: 'libusb_set_debug' is deprecated [-Wdeprecated-declarations]
usb/lowlevel/libusb/c/libusb.h:1351:1: note: 'libusb_set_debug' has been explicitly marked deprecated here
usb/lowlevel/libusb/c/libusb.h:70:50: note: expanded from macro 'LIBUSB_DEPRECATED_FOR'
docker run -it --rm \
		-v /Users/tomassusanka/go/src/github.com/trezor/trezord-go:/trezord \
		-w /trezord \
		-e CGO_ENABLED=1 \
		--entrypoint /trezord/release/macos/build-fat.sh \
		docker.elastic.co/beats-dev/golang-crossbuild:1.19-darwin-arm64-debian10
+ /usr/local/osxcross/bin/lipo -create release/macos/build/trezord-arm64 release/macos/build/trezord-amd64 -output release/macos/build/trezord
fatal error: /usr/local/osxcross/bin/lipo: can't open input file: release/macos/build/trezord-amd64 (No such file or directory)
make: *** [.binary] Error 1

```

</details>

igor-hnizdo commented:
Ahh it seems like that, yes. Seems like the base image is arm... hmmm.

tsusanka commented:
I am not saying it is a must-have. If we have working builds in CI I am fine with that. But it would be nice.

igor-hnizdo commented:
Ah no, the error was because of the `.package` step, which is still not really fixed.

Revert your local changes and run just `make .binary`, that should work now. I see that it worked in your log, and failed on the .package step.

I am now fixing .package step.

tsusanka commented:
Oh right 🤦 . Great, that works 🎉 . So let's wait for the other fixes or do you want to make a new PR?

igor-hnizdo commented:
I can add it here yeah

igor-hnizdo commented:
I fixed the release stuff now too.

There are several issues that needed to be fixed:

* we need to do some changes to make the build workable on Mac docker, as Mac FS is case insensitiv
e
* we need to update Fedora, as the old one does not have ARM base image
* xar and bomutils have old source code that is failing on ARM and on new Fedora. The projects are
no longer maintained; I have added PRs to the most popular forks, but still I made my own forks.

I can instead pull the code into monorepo here, I don't know what is preferable.

igor-hnizdo commented:
Note that notarization still does not work.

There is this project that implements Apple notarization in Rust, I think it will work. I need the JSON keys though :) let's discuss this off-band

https://gregoryszorc.com/docs/apple-codesign/main/index.html

https://gregoryszorc.com/docs/apple-codesign/main/apple_codesign_getting_started.html#obtaining-a-code-signing-certificate

tsusanka commented:
> xar and bomutils have old source code that is failing on ARM and on new Fedora. The projects are
no longer maintained; I have added PRs to the most popular forks, but still I made my own forks.

Is there any chance replacing these with something else? Both have last commits in 2014, and patching such old libraries seems crazy. Isn't there some new alternative? Forking and patching these is a slippery slope..

tsusanka commented:
Now I see your fork of bomutils is fork of another fork which has some additional fixes. Let's wait if they accept your PR https://github.com/BC-SECURITY/bomutils/pull/2. So the question above applies for `xar` mainly.

tsusanka commented:
Btw we use `xar` for producing the .pkg package installer, correct? Can we replace that with .dmg? Because we distribute Suite like that, and I think it works just fine.

igor-hnizdo commented:
I don't know.

I don't think dmg can register itself as launch agent and start on login. It's much more work :) this was easier.

Yeah it's to make the .pkg

igor-hnizdo commented:
I made a similar PR here

https://github.com/tpoechtrager/xar/pull/1

igor-hnizdo commented:
Anyway, I fixed it by using `CFLAGS` that set char as signed.

It seems that C standard does not define whether char is by default signed or unsigned, and it defaults as unsigned on ARM for performance reasons.

Fun with C! 

Anyway yeah it works now. I don't think I will rework this to DMG, sorry. :( I don't want to debug how to auto-start trezord with that setup.

tsusanka commented:
> Anyway yeah it works now. I don't think I will rework this to DMG, sorry. :( I don't want to debug how to auto-start trezord with that setup.

Yeah sure, don't worry, that was just an idea how to get rid off the forks, now we do not need that :). Let me have a look!

tsusanka commented on release/macos/Dockerfile:
```diff
@@ -10,13 +10,10 @@ RUN dnf update -y
-RUN git clone https://github.com/mackyle/xar.git
-RUN cd xar && cd xar && ./autogen.sh --noconfigure && ./configure && make && make install
+RUN git clone https://github.com/tpoechtrager/xar
+RUN cd xar && cd xar && ./autogen.sh --noconfigure && CFLAGS="-fsigned-char" ./configure && make && make install
```
Newline pls 😇

tsusanka reviewed (approved):
Tested and works 🎉. One tiny nit-pick.

tsusanka commented:
I would merge then and let's do the notarization and windows/linux fixes in another PRs.
